rust/otel-arrow-rust/src/encode/record/array.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides adaptive array builders for Arrow arrays.
+//!
+//! Often for OTel-Arrow, we have columns that are optional on the schema. For example, the Boolean
+//! column may not be present in a record batch representing a list of attributes only be present
+//! if there are some boolean type values in the list.
+//!
+//! There are also cases where we want to dynamically use dictionary encoding with the smallest index
+//! the cardinality of the allows.
+//!
+//! This module contains adaptive array builders that can dynamically create either no array (for
+//! an all-null) column, an array that may be a dictionary, or an array or native types. It will
+//! handle converting between different builders dynamically  based on the data which is appended.
+//! ```
+
+use arrow::array::{ArrayRef, StringBuilder, StringDictionaryBuilder};
+use arrow::datatypes::{DataType, UInt8Type, UInt16Type};
+
+use crate::arrays::NullableArrayAccessor;
+
+use dictionary::{
+    AdaptiveDictionaryBuilder, ConvertToNativeHelper, DictionaryArrayBuilder,
+    DictionaryBuilderError, DictionaryOptions, UpdateDictionaryIndexInto,
+};
+
+pub mod dictionary;
+pub mod string;
+
+/// This is the base trait that array builders should implement.
+pub trait ArrayBuilder {
+    type Native;
+
+    fn append_value(&mut self, value: &Self::Native);
```

Comment:
I’m not sure the signature of this function will be sufficient. You will probably need to include the `current_row` as a parameter. For example, if all previous values were null and the new value is non-null, this would trigger the actual creation of the underlying array. In such a case, you’d need to set indices `0..current_row - 1` to null and then insert the current value at `current_row`.

The other option is to keep track of all previously added null values. I’m not sure which of these approaches would be more efficient.

Reply:
FYI opened a PR on arrow-rs to add the `add_nulls` method to the other types of builders we'll use in this case 
https://github.com/apache/arrow-rs/pull/7606